include/Analysis/RotationAnalysis/RotationAnalysis.h

```diff
@@ -1,215 +1,265 @@
-  // state.
-  void clear() {
-    accessedIndices.clear();
-    status = Status::Uninitialized;
-  }
-
   bool empty() const { return accessedIndices.empty(); }
 
-  bool isOverdetermined() const { return status == Status::Overdetermined; }
-
-  bool isUninitialized() const { return status == Status::Uninitialized; }
-
   void addRotation(int64_t index) { accessedIndices.insert(index); }
 
-  bool operator==(const RotationSets &rhs) const {
-    return tensor == rhs.tensor && status == rhs.status &&
-           accessedIndices == rhs.accessedIndices;
+  // Returns true if the accessed indices constitute all indices of the reduced
+  // tensor.
+  bool isComplete() const {
+    auto tensorType = tensor.getType().dyn_cast<RankedTensorType>();
+    assert(tensorType &&
+           "Internal state of RotationAnalysis is broken; tensor must have a "
+           "ranked tensor type");
+
+    // std::set is ordered, so min/max is first/last element of the set
+    int64_t minIndex = *accessedIndices.begin();
+    int64_t maxIndex = *accessedIndices.rbegin();
+    return minIndex == 0 && maxIndex == tensorType.getShape()[0] - 1 &&
+           accessedIndices.size() == tensorType.getShape()[0];
   }
 
-  const std::unordered_set<int64_t> &getAccessedIndices() const {
+  const std::set<int64_t> &getAccessedIndices() const {
     return accessedIndices;
   }
 
   Value getTensor() const { return tensor; }
 
+  Value getRoot() const { return root; }
+
   void print(raw_ostream &os) const {
-    os << tensor << ": [";
+    os << "{ opName: " << (opName.has_value() ? opName->getStringRef() : "None")
+       << "; " << " tensor: " << tensor << "; " << "rotations: [";
     for (auto index : accessedIndices) {
       os << index << ", ";
     }
-    os << "]";
+    os << "]; root: " << root << "; }";
   }
 
-  static RotationSets overdetermined() {
-    RotationSets sets;
-    sets.status = Status::Overdetermined;
-    return sets;
-  }
-
-  static RotationSets from(Value tensor) {
-    RotationSets sets;
-    if (!tensor.getType().isa<RankedTensorType>()) {
-      sets.status = Status::Uninitialized;
-      return sets;
-    }
-
-    sets.status = Status::Normal;
-    sets.tensor = tensor;
-    if (auto blockArg = dyn_cast<BlockArgument>(tensor)) {
-      sets.addRotation(0);
-    }
-    return sets;
+  // Construct a "leaf" of a reduction, i.e., a PartialReduction that represents
+  // no operations applied to a starting tensor SSA value.
+  static PartialReduction initializeFromValue(Value tensor) {
+    PartialReduction reduction;
+    reduction.tensor = tensor;
+    reduction.root = tensor;
+    reduction.opName = std::nullopt;
+    // In the FHE world, the only extractible element (without a rotation) of a
+    // packed ciphertext is the constant term, i.e., the first element of the
+    // tensor. So a tensor by itself is always considered a reduction by that
+    // first element.
+    reduction.addRotation(0);
```

Comment:
I was a little confused about this one and went back and forth trying to understand it. If there are no operations on it at this point, why would it be assumed to be accessed at 0? for e.g. what if the only op is a tensor.rotate %1 and then a tensor.extract?

Reply:
The way I understand it (which may not be entirely correct) is that scalars in BGV/BFV/CKKS are interpreted as lying in the constant term of a ciphertext, and is zero-padded to fill up the rest of the tensor. Similarly, if we weren't aligning rotations, then the "default" thing to do to get a scalar addition of two values in a tensor is rotate them both to the zero-th slot and SIMD-add there. I think this is more of a convention than a deduction, but with that mental model, if the scalar you want is already in slot 0 then doing nothing is the same as accessing that index.

> what if the only op is a tensor.rotate %1 and then a tensor.extract?

I think this would be incorrect if you think of this analysis as "what indices are accessed by ANY op ending at a particular SSA value," but we're restricting it to reduction ops by specific binary operations.

More pragmatically, rotations by zero are deleted by the canonicalizer, and because they're missing from the IR, the analysis can't recover them and improperly marks a full reduction as incomplete. If I removed this and had the check assume zero is always included (not sure what the alternative is), that would be incorrect if the original untouched tensor is not part of the reduction (sum all but the 0th entry).